### ISSUE TITLE
learning rate schedule for different number of data

### ISSUE THREAD

**Issue by kwea123** [NVIDIA/semantic-segmentation#30]
Do you use the same learning rate schedule for experiments with and without label propagation?
For example, the original training set (N=2975 images), the sdcaug+-3 (3N images) and the sdcaug +-3 accumulated (7N images), do they share the same learning rate schedule?

I'm trying to reproduce your results, I'm currently using the exact same learning rate schedule for all the above cases, and I observe that the validation loss of the +-3 accumulated experiment saturates very quickly, and surpasses the losses of the original one and the +-3 cases at about the 10th epoch already. Surprisingly, its mean IoU still goes up; but doesn't the high val loss mean that it's overfitting already?

The following graphs are the mean IoU and val loss, `pm3` means +-3 and `pm3all` means +-3 accumulated. The x-axis is steps (or epochs)
![a](https://user-images.githubusercontent.com/11364490/66110886-aa1f1e80-e602-11e9-85c6-42451cc65078.png)
![b](https://user-images.githubusercontent.com/11364490/66110891-abe8e200-e602-11e9-99ce-9dd049a00419.png)

I'm suspecting that using the same learning rate schedule makes the learning rate too high for the accumulated case: when 7N images are trained, the original case has already decreased for some ![lambda^7](https://user-images.githubusercontent.com/11364490/66111291-84dee000-e603-11e9-9734-76f9e83aa68f.gif), but the accumulated case is decreased for only once. The same phenomenon should apply to +-3 case, but it doesn't overfit, maybe it's because the learning rate is still not too high for it.

Perhaps a better way for training +-3 accumulated case is to set a lower learning rate at the beginning, or to decrease the learning rate more (e.g. decrease by ![lambda^7](https://user-images.githubusercontent.com/11364490/66111291-84dee000-e603-11e9-9734-76f9e83aa68f.gif)?)

What's your opinion?

**Comment by bryanyzhu** [comment]
Great observations. 

1. High val loss maybe an indicator of overfitting. But you need to evaluate the model on validation or test set to see if they are actually overfitting. 

2. We didn't try different learning rate schedules because we want to make fair comparison (keep hyper-parameters the same). But yes, you are making a good point. A smaller learning rate may be more suitable for accumulated case.

**Comment by kwea123** [comment]
By the way, what is the epoch that the model has the best performance? By default it trains 175 epochs but the best performance is achieved for epoch 120 for me. Is that normal?

**Comment by bryanyzhu** [comment]
It is normal, I had a model before that achieves best performance at epoch 96. 

Semantic segmentation training is not as stable as other tasks. That's why in our paper, we perform several same runs and report the mean and std of the performance, instead of just reporting the best.

**Comment by kwea123** [comment]
May I also ask, why is the learning rate has this kind of shape? More precisely, why do you increase it again to 1e-3 after it reaches about 4e-4?
![image](https://user-images.githubusercontent.com/11364490/66453565-c35f1980-ea9f-11e9-82ea-4b818204159c.png)

**Comment by bryanyzhu** [comment]
The learning rate has this kind of shape because of our modified optimizer, which can be seen [here](https://github.com/NVIDIA/semantic-segmentation/blob/master/optimizer.py#L35).

The intuition is once we turn off the boundary relaxation loss (at epoch 100), we want the model to re-focus on the boundary again and try to achieve higher mIoU. But if the learning rate is too low, it may be hard for the model to restart learning. So we manually increase the learning rate, hoping the model can learn new stuff. 

@karansapra Correct me if I'm wrong. And add more comments if you like. Thank you.

**Comment by kwea123** [comment]
Did you do any comparison of this schedule and a normal one e.g. which keeps decreasing the learning rate?

**Comment by bryanyzhu** [comment]
Yes, it should help. But I didn't do this kind of comparison.

@karansapra Any experience on this? Thanks.
